### DISCUSSION TITLE
Why aren't non-existent functions detected at compile time?

### DISCUSSION THREAD

**Discussion by TheYellowArchitect** [godotengine/godot-proposals#5549] [Engine Core]
I have to ask this after https://github.com/godotengine/godot/issues/66943
Look at this example
```python
@export var bullet_scene : PackedScene

func _ready():
    var bullet_instance = bullet_scene.instance()
```
Why does it even compile? It's obvious that it will crash/error on runtime, the function `.instance` does not exist. `.instantiate` does, and works.

While I am curious to know, with my limited knowledge of Godot I do think it's a bug (python works like this but python isn't in this project), but I want to confirm hence I ask. After all, static compilation does happen, but I guess some of it is offloaded to runtime?

**Comment by Calinou** [comment]
I don't think it's a bug – it's just a missing static check. Implementing this for built-in types may be possible, but not trivial.

**Reply by TheYellowArchitect** [reply]
Thank you for your answer. I believe you that it's hard to implement (is there an issue created for this?), just wanted to confirm its not a bug, as it was annoying

Off-topic: idk how github discussions work. Is there a way to "close" this "issue"?

**Reply by Calinou** [reply]
> (is there an issue created for this?)

I don't remember seeing one. We only have a few contributors knowledgeable with working on GDScript itself, so this may take a while to be implemented (if ever).

> Off-topic: idk how github discussions work. Is there a way to "close" this "issue"?

Discussions are never closed, unlike issues.

**Comment by dalexeev** [comment]
> Why does it even compile? It's obvious that it will crash/error on runtime, the function `.instance` does not exist.

In theory, you can assign to `bullet_scene` a `PackedScene` instance with an attached script that implements the `instance` function.
Or, which is almost the same, `PackedScene` _may have_ a child class with `instance` function.

The following code is also valid, although not type-safe:

```gdscript
var node: Node = Label.new() # OK, Label inherits Node.
node.name = "MyNode" # OK, type-safe. Node.name exists.
node.text = "test" # OK, but not type-safe. Node.text does not exist, but Label.text exists.
node.lalala = 123 # OK, but not type-safe. Runtime error.
```

If you want more strict checks, you can change these settings:

![](https://user-images.githubusercontent.com/47700418/194314049-020b0e5b-d0dd-4034-9d87-bebef78bf191.png)

**Reply by TheYellowArchitect** [reply]
Godlike reply. I didn't think of any of these use-cases. It seems like... its not a bug, its a feature hahaha

**Reply by TheYellowArchitect** [reply]
Being a newbie with Godot, trying to learn the API, I will definitely make use of those project settings options (didn't know they existed)
Thanks! <3
